### ISSUE TITLE
Support tool-bearing requests on target-verified MTP without weakening stateful processors

### ISSUE THREAD

**Issue by korellas** [raullenchai/vllm-mlx#2886]
## What are you trying to do?

Run OpenAI-compatible coding or automation agents with function tools while retaining target-verified MTP acceleration. Tool schemas, structured tool output, and loop protection must remain enabled; selecting MTP should accelerate eligible tool-bearing requests rather than requiring tools to be removed.

## What's blocking you today?

Every request with `tools` installs `AgentRepetitionLogitsProcessor`. Rapid-MLX classifies that processor as custom/stateful rather than MTP-safe, so the request stays on ordinary decode:

```text
[MTP-vendored] uid=0 remains on plain decode: custom or stateful logits processor is active (active=1 [AgentRepetitionLogitsProcessor], safe=0 [-]).
```

This means an MTP-enabled Qwen3.8-27B server receives no speculative-decoding benefit for normal agent requests, even when no grammar or reasoning-budget processor is active. Issue #2658 identified the user-visible limitation, and #2683 made the fallback visible in Desktop, but the server capability itself remains unavailable.

## Current workaround (if any)

Run Rapid-MLX with `--no-spec-decode` for tool workloads, disable tools to use MTP, or use a separate runtime whose MTP verifier supports tool constraints. The first two options give up either the requested acceleration or the required agent capability.

## Proposed approach (optional)

Start with an explicit speculative transaction contract for `AgentRepetitionLogitsProcessor`, without treating arbitrary processors as safe:

- evaluate the processor at every proposed position against committed history plus the tentative draft prefix;
- commit processor/history state only for target-accepted tokens;
- roll back rejected, cancelled, or aborted speculative suffixes;
- permit the existing MTP path only when that contract is available;
- keep grammar, reasoning-budget, suppression, seeded-RNG, and unknown custom processors fail-closed until each has its own verified contract.

This narrower first step would cover ordinary `tools` requests that currently contain only the agent repetition guard, while preserving the existing correctness boundary for other stateful processors.

## Alternatives you considered

- Removing the repetition guard from tool requests would expose the exact agent loops it was added to prevent.
- Marking the processor safe without per-position tentative history and rollback could change the target distribution after a rejected draft.
- Disabling tools or MTP is operationally reliable but defeats one of the two requested capabilities.
- Other MLX runtimes demonstrate that tool-aware MTP is feasible when constraints and loop protection are integrated into target verification, so this is an implementation boundary rather than a fundamental MTP limitation.

## Hardware (if relevant)

Mac Studio, Apple M3 Ultra, 256 GB

## Model (if relevant)

Qwen3.8-27B 8-bit MLX checkpoint with a matching 8-bit MTP sidecar

## Are you willing to contribute this?

No, I'm just requesting

## Additional context

- #2655 added request-local sampled MTP and explicitly retained ordinary decode for stateful tool processors.
- #2658 tracked the resulting tools-enabled fallback.
- #2683 resolved the visibility/UI portion by reporting `MTP paused`; this request is specifically for the remaining engine capability.

Acceptance should include a real tool-bearing request proving nonzero MTP proposals/accepts, valid structured tool output, unchanged loop protection, and correct commit/rollback behavior at K > 1 and on cancellation.

**Comment by korellas** [comment]
Apologies — I submitted this through GitHub CLI. The CLI can reproduce the form fields in Markdown, but it cannot submit the repository’s YAML Issue Form itself, so the template label was not applied and the spam guard flagged the issue. Sorry for bypassing the intended submission path.

**Comment by korellas** [comment]
Related prior MTP work:

- [#2654 — Desktop MTP silently falls back under default sampling](https://github.com/raullenchai/Rapid-MLX/issues/2654): established the request-eligibility problem for ordinary sampled requests.
- [#2655 — serve Desktop default sampling without silent fallback](https://github.com/raullenchai/Rapid-MLX/pull/2655): enabled target-verified sampled MTP while explicitly keeping stateful tool processors on plain decode.
- [#2658 — Desktop default tools keep MTP-enabled chats on ordinary decode](https://github.com/raullenchai/Rapid-MLX/issues/2658): tracked either making the limitation visible or adding a safe processor transaction contract.
- [#2683 — show per-request MTP availability](https://github.com/raullenchai/Rapid-MLX/pull/2683): resolved the visibility/UI portion only; enabling speculative execution for stateful processors was explicitly out of scope.

**Comment by raullenchai** [comment]
**Closing as completed — the engine capability landed in #3032.**

Phase-0 audit (per the issue-implementation portfolio) maps the acceptance criteria against current `main` (`7f34f5d2c`):

- ✅ **Per-position tentative evaluation.** `AgentRepetitionLogitsProcessor.mtp_apply(tentative_token_ids, logits)` evaluates at every proposed position against committed scheduler-owned history plus the accepted tentative prefix (`vllm_mlx/repetition_guard.py`). No supported loop can arm below 48 output tokens with production thresholds.
- ✅ **Commit only on target acceptance.** `mtp_snapshot_state()` / `mtp_restore_state()` capture and restore `interventions`, `last_match`, and `_last_intervention_length`; `spec_decode/mtp/generator.py` commits guard state only as tokens are delivered and rolls back on rejection, cancellation, processor raise, or abort. A rejected draft never consumes an intervention or poisons the next ordinary token.
- ✅ **Ordinary decode preserves the contract** — the committed-output path is unchanged for non-MTP decode.
- ✅ **Eligibility is narrow.** Only processors exposing the three `mtp_*` methods are routed through the transactional path (`generator.py`); grammar, reasoning-budget, suppression, seeded-RNG, and unknown processors remain on plain decode, fail-closed.
- ✅ **Metrics/evidence + tests.** `test_repetition_guard.py`, `test_mtp_spec_decode.py`, and `test_speculative_request_policy.py` cover reject-first, partial accept at K>1, cancellation rollback, tool-guard commit-on-deliver, restore-on-raised-target, and ordinary-decode parity — 165 passing.

Noted gap (non-blocking, hardening only): eligibility is currently duck-typed via `callable(getattr(…))` rather than an explicit capability allowlist. The functional acceptance is met and the issue remains closed; a follow-up may tighten that mechanism if desired.

Tracked as resolved by #3032. Closing.
